components/archiving/projectController.js

```diff
@@ -0,0 +1,436 @@
+      } else if (query == urls.filesUnloaded) {
+        projects = await archiveDao.getByFilesNotUnloaded(projectEntityType);
+        res.status(200).send(replaceGenericKeyName(projects));
+      }
+    } else {
+      projects = await archiveDao.streamAllProjects((stream) => {
+        res.status(200);
+        stream
+          .pipe(createProjectKeyReplacementStream())
+          .pipe(JSONStream.stringify())
+          .pipe(res);
+        stream.on('error', (err) => {
+          // log the error and prematurely end the response
+          logger.error(err);
+          res.end();
+        });
+      });
+      logger.info({
+        streamRowsProcessedProjects: projects.processed,
+        streamingDurationProjects: projects.duration,
+        method: '',
+      });
+    }
+    next();
+  } catch (e) {
+    handleErrors(e, 'Error getting projects', logger, next);
+  }
+};
+
+const isCompletelyArchived = (kase) => {
+  return (
+    kase.filesCopiedToOffsiteArchiveStagingDir != null &&
+    kase.commvaultBackupJobId != null &&
+    kase.filesLoadedIntoVidarrArchival != null &&
+    kase.filesUnloaded != null
+  );
+};
+
+const hasArchivingStarted = (kase) => {
+  return kase.filesCopiedToOffsiteArchiveStagingDir != null;
+};
+
+const addProjectArchive = async (req, res, next) => {
+  try {
+    //authenticate api-key from header before continuing
+    await authenticator.authenticateRequest(req);
+
+    const existingProjects = await archiveDao.getByArchiveEntityIdentifier(
+      req.body.projectIdentifier,
+      false,
+      projectEntityType
+    );
+    if (existingProjects == null || !existingProjects.length) {
+      await upsert(req.body, true);
+      res.status(201).end();
+    } else {
+      for (let existingProject of existingProjects) {
+        // check for conflicting changes
+        let errors = [];
+        let projectStoppedProcessing = false;
+        if (existingProject.requisitionId != req.body.requisitionId) {
+          errors.push(
+            `Requisition (${req.body.requisitionId}) from request does not match requisition ${existingProject.requisitionId} for project ${req.body.projectIdentifier}`
+          );
+          projectStoppedProcessing = true;
+        }
+        if (!arraysEquals(existingProject.limsIds, req.body.limsIds)) {
+          errors.push(
+            `LIMS IDs (${req.body.limsIds}) from request do not match LIMS IDs ${existingProject.limsIds} for project ${req.body.projectIdentifier}`
+          );
+          projectStoppedProcessing = true;
+        }
+        if (
+          !arraysEquals(
+            existingProject.workflowRunIdsForOffsiteArchive,
+            req.body.workflowRunIdsForOffsiteArchive
+          )
+        ) {
+          errors.push(
+            `Requested offsite archive files list ${req.body.workflowRunIdsForOffsiteArchive} does not match offsite files archive list ${existingProject.workflowRunIdsForOffsiteArchive} for project ${existingProject.entityIdentifier}`
```

Comment:
This has the potential to be an extremely long error message. Would it make sense to change how we compare array equality to return any discordances, which could then be used to make more informative error messages?